docarray/doc_index/backends/hnswlib_doc_index.py

```diff
@@ -0,0 +1,400 @@
+        self._db_config = cast(HnswDocumentIndex.DBConfig, self._db_config)
+        self._work_dir = self._db_config.work_dir
+        load_existing = os.path.exists(self._work_dir) and os.listdir(self._work_dir)
+        Path(self._work_dir).mkdir(parents=True, exist_ok=True)
+
+        # HNSWLib setup
+        self._index_construct_params = ('space', 'dim')
+        self._index_init_params = (
+            'max_elements',
+            'ef_construction',
+            'M',
+            'allow_replace_deleted',
+        )
+
+        self._hnsw_locations = {
+            col_name: os.path.join(self._work_dir, f'{col_name}.bin')
+            for col_name, col in self._columns.items()
+            if col.config
+        }
+        self._hnsw_indices = {}
+        for col_name, col in self._columns.items():
+            if not col.config:
+                continue  # do not create column index if no config is given
+            if load_existing:
+                self._hnsw_indices[col_name] = self._load_index(col_name, col)
+            else:
+                self._hnsw_indices[col_name] = self._create_index(col)
+
+        # SQLite setup
+        self._sqlite_db_path = os.path.join(self._work_dir, 'docs_sqlite.db')
+        self._sqlite_conn = sqlite3.connect(self._sqlite_db_path)
+        self._sqlite_cursor = self._sqlite_conn.cursor()
+        self._create_docs_table()
+        self._sqlite_conn.commit()
+
+    ###############################################
+    # Inner classes for query builder and configs #
+    ###############################################
+    class QueryBuilder(BaseDocumentIndex.QueryBuilder):
+        def build(self, *args, **kwargs) -> Any:
+            return self._queries
+
+    @dataclass
+    class DBConfig(BaseDocumentIndex.DBConfig):
+        work_dir: str = '.'
+
+    @dataclass
+    class RuntimeConfig(BaseDocumentIndex.RuntimeConfig):
+        default_column_config: Dict[Type, Dict[str, Any]] = field(
+            default_factory=lambda: {
+                np.ndarray: {
+                    'dim': 128,
+                    'space': 'l2',  # 'l2', 'ip', 'cosine'
+                    'max_elements': 1024,
+                    'ef_construction': 200,
+                    'ef': 10,
+                    'M': 16,
+                    'allow_replace_deleted': True,
+                    'num_threads': 1,
+                },
+                # `None` is not a Type, but we allow it here anyway
+                None: {},  # type: ignore
+            }
+        )
+
+    ###############################################
+    # Implementation of abstract methods          #
+    ###############################################
+
+    def python_type_to_db_type(self, python_type: Type) -> Any:
+        """Map python type to database type."""
+        for allowed_type in HNSWLIB_PY_VEC_TYPES:
+            if issubclass(python_type, allowed_type):
+                return np.ndarray
```

Comment:
return `np.ndarray` or `allowed_type `

Reply:
I don't think that implements the same logic